### ISSUE TITLE
Unable to add a certificate credential to an Azure AD application or service principal by referencing a certificate stored in Azure Key Vault

### ISSUE THREAD

**Issue by CaptainStealthy** [hashicorp/terraform-provider-azuread#385]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritise this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritise the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureAD Provider) Version

```
Terraform v0.13.5
+ provider registry.terraform.io/hashicorp/azuread v1.2.2
+ provider registry.terraform.io/hashicorp/azurerm v2.43.0
```

### Affected Resource(s)

* `azuread_application_certificate`

### Terraform Configuration Files

```hcl
resource "azurerm_key_vault_certificate" "cert" {
  name         = var.resourcegroup_name
  key_vault_id = azurerm_key_vault.kv.id

  certificate_policy {
    issuer_parameters {
      name = "Self"
    }

    key_properties {
      exportable = true
      key_size   = 2048
      key_type   = "RSA"
      reuse_key  = true
    }

    lifetime_action {
      action {
        action_type = "EmailContacts"
      }

      trigger {
        days_before_expiry = 30
      }
    }

    secret_properties {
      content_type = "application/x-pkcs12"
    }

    x509_certificate_properties {
      # Server Authentication = 1.3.6.1.5.5.7.3.1
      # Client Authentication = 1.3.6.1.5.5.7.3.2
      extended_key_usage = ["1.3.6.1.5.5.7.3.1"]

      key_usage = [
        "cRLSign",
        "dataEncipherment",
        "digitalSignature",
        "keyAgreement",
        "keyCertSign",
        "keyEncipherment",
      ]

      subject            = "CN=${var.resourcegroup_name}"
      validity_in_months = 12
    }
  }
}

resource "azuread_application" "app" {
  display_name            = join("", [var.resourcegroup_name, "-ManagedByTerraform"])
  homepage                = local.api_url
  owners                  = concat(data.azuread_users.app_owners.object_ids, [local.azurerm_principal_id])
  prevent_duplicate_names = true
}

resource "azuread_service_principal" "app" {
  application_id = azuread_application.app.application_id
}

resource "azuread_application_certificate" "app_cert" {
  application_object_id = azuread_application.app.id
  type                  = "AsymmetricX509Cert"
  value                 = azurerm_key_vault_certificate.cert.certificate_data
  end_date              = azurerm_key_vault_certificate.cert.certificate_attribute[0].expires
  start_date            = azurerm_key_vault_certificate.cert.certificate_attribute[0].not_before
}
```

### Expected Behavior

Expected TF to add a certificate credential to the AAD application by pulling the newly created PKCS12 certificate in Azure Key Vault.

### Actual Behavior

Deployment failed with the following error:

```
azuread_application_certificate.app_cert: Creating...

Error: Creating certificate credentials "<redacted>" for application with object ID "<redacted>"

graphrbac.ApplicationsClient#UpdateKeyCredentials: Failure responding to
request: StatusCode=400 -- Original Error: autorest/azure: Service returned an
error. Status=400 Code="Unknown" Message="Unknown service error"
Details=[{"odata.error":{"code":"Request_BadRequest","date":"2021-01-19T01:12:10","message":{"lang":"en","value":"Invalid
certificate: Key value is invalid
certificate"},"requestId":"<redacted>","values":[{"item":"PropertyName","value":"keyCredentials.value"},{"item":"PropertyErrorCode","value":"InvalidKeyValue"}]}}]

```

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

I tried changing the credential type from `AsymmetricX509Cert` to `Symmetric` and got this:

```
Encrypted secret cannot be empty and can be at most 1024 bytes. Current length is 1859
```

**Comment by CaptainStealthy** [comment]
Just in case anyone else is facing the same problem and happens to stumble upon this thread, I thought I should provide an update. Upon several hours of Google-fu...

It appears that this is actually an underlying issue in the `azurerm` provider:
* https://github.com/terraform-providers/terraform-provider-azurerm/issues/8072
* https://github.com/terraform-providers/terraform-provider-azurerm/issues/8482
* https://github.com/terraform-providers/terraform-provider-azurerm/pull/8184

Basically, the `azurerm_keyvault_certificate` resource outputs certs in hexadecimal format, which can't be read/converted by this provider - or that's how I interpret things, anyway.

I am going to see if I can come up with an alternative workaround while waiting for the aforementioned PR to be closed. Hope to report back soon!

**Comment by CaptainStealthy** [comment]
Thanks to help (directly and indirectly) from @manicminer, @matti, and @akingscote (ref - https://github.com/hashicorp/terraform/issues/26163#issuecomment-688872272), I've got myself a workaround! 😁 

This little snippet will take the certificate (which is in hexadecimal) from Azure Key Vault and convert it into a base64 encoded string, which can then be fed into `azuread_application_certificate` or `azuread_service_principal_certificate`.

manicminer, per our Slack conversation, I am leaving this open for now. I should note, the root cause isn't in this provider - a comment in [#8482 of `azurerm`](https://github.com/terraform-providers/terraform-provider-azurerm/issues/8482) suggests that the hex attribute is going to be removed in a future release in favor of just returning a base64 string, so I wouldn't put it too terribly high on your priority list.

Thanks again, all!

```
# This code requires other pieces in order to function, obviously
# But this is the specific code needed to take hex certificate returned by azurerm and convert it into a base64 string

module "cert-hex-decode" {
  source  = "matti/resource/shell"
  version = "1.1.0"
  command = "echo ${azurerm_key_vault_certificate.cert.certificate_data} | xxd -r -p | base64 -w 0"
  trigger = azurerm_key_vault_certificate.cert.version
}

resource "azuread_application_certificate" "app_cert" {
  application_object_id = azuread_application.app.id
  type                  = "AsymmetricX509Cert"
  value                 = module.cert-hex-decode.stdout
  end_date              = azurerm_key_vault_certificate.cert.certificate_attribute[0].expires
  start_date            = azurerm_key_vault_certificate.cert.certificate_attribute[0].not_before
  depends_on            = [module.cert-hex-decode.id] # Not sure if this is needed, but I put it in for sanity's sake
}
```

**Comment by manicminer** [comment]
Thanks for raising this @CaptainStealthy. I agree it would be good for the azurerm_key_vault_certificate resource to export a PEM encoded certificate, however I believe this is still a useful feature here since certificate data can come from a variety of sources.
